include/envoy/http/hash_policy.h

```diff
@@ -11,7 +11,7 @@ namespace Http {
 
 class Hashable {
 public:
-  virtual absl::optional<uint64_t> hash() const { return absl::nullopt; };
+  virtual absl::optional<uint64_t> hash() const PURE;
```

Comment:
Does it still need to be optional as well?

Reply:
I want to keep it "relaxed", in the case when there is no data available (say, header not available yet), return the same value as the evaluate() suggests.
We can definitely require subclasses of the Hashable class to return 0, but optional maybe aligns better with the existing interface?